EIPS/eip-55.md

```diff
@@ -13,25 +13,47 @@ created: 2016-01-14
 Code:
 
 ``` python
-from ethereum import utils
+import eth_utils
+
 
 def checksum_encode(addr): # Takes a 20-byte binary address as input
-    o = ''
-    v = utils.big_endian_to_int(utils.keccak256(addr.hex()))
-    for i, c in enumerate(addr.hex()):
-        if c in '0123456789':
-            o += c
+    hex_addr = addr.hex()
+    checksummed_buffer = ""
+
+    # Treat the hex address as ascii/utf-8 for keccak256 hashing
+    hashed_address = eth_utils.keccak(text=hex_addr).hex()
+
+    # Iterate over each character in the hex address
+    for nibble_index, character in enumerate(hex_addr):
+
+        if character in "0123456789":
+            # We can't upper-case the decimal digits
+            checksummed_buffer += character
+        elif character in "abcdef":
+            # Check if the corresponding hex digit (nibble) in the hash is 8 or higher
+            hashed_address_nibble = int(hashed_address[nibble_index], 16)
+            if hashed_address_nibble > 7:
+                checksummed_buffer += character.upper()
+            else:
+                checksummed_buffer += character
         else:
-            o += c.upper() if (v & (2**(255 - 4*i))) else c.lower()
-    return '0x'+o
+            raise eth_utils.ValidationError(
+                f"Unrecognized hex character {character!r} at position {nibble_index}"
+            )
```

Comment:
Not a fan of breaking this line into 3 lines to conform to a ridiculously out of date standard, but beggars can't be choosers so I'll accept.  😊